### PR TITLE
Move to PrecompileTools

### LINKED ISSUE #1182: Precompile failures on julia nightly
```
Failed to precompile Flux [587475ba-b771-5e3f-ad9e-33799f191a9c] to "/home/runner/.julia/compiled/v1.11/Flux/jl_hOZ50V".
ERROR: LoadError: AssertionError: precompile(Tuple{typeof((which(__broadcast, (Any, Size, Tuple{Vararg{Size}}, Vararg{Any}))).generator.gen), Any, Any, Any, Any, Any, Any})
Stacktrace:
 [1] _precompile_()
   @ StaticArrays ~/.julia/packages/StaticArrays/t0DdQ/src/precompile.jl:25
 [2] top-level scope
   @ ~/.julia/packages/StaticArrays/t0DdQ/src/StaticArrays.jl:139
 [3] include(mod::Module, _path::String)
   @ Base ./Base.jl:489 [inlined]
 [4] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{…}, dl_load_path::Vector{…}, load_path::Vector{…}, concrete_deps::Vector{…}, source::String)
   @ Base ./loading.jl:2216
 [5] top-level scope
   @ stdin:3
Some type information was truncated. Use `show(err)` to see complete types.
in expression starting at /home/runner/.julia/packages/StaticArrays/t0DdQ/src/StaticArrays.jl:1
in expression starting at stdin:3
ERROR: LoadError: Failed to precompile StaticArrays [90137ffa-7385-5640-81b9-e5203721[818](https://github.com/r3tex/ObjectDetector.jl/actions/runs/5577159290/jobs/10189481031?pr=88#step:4:822)2] to "/home/runner/.julia/compiled/v1.11/StaticArrays/jl_HtO2XA".
```

### LINKED ISSUE #1184: Move to PrecompileTools.jl ?
There are a few hardcoded precompile directives, 3 of which just broke on julia v1.11 (#1182) and disabled in #1183 

Perhaps PrecompileTools should be used instead?

### PR DESCRIPTION
**Pull request by IanButterworth**
Fixes #1184
Fixes #1182
Closes #1183

@timholy would you mind helping me figure out the commented out ones?

### CODE REVIEW

File: src/precompile.jl [hunk 1]
```diff
@@ -1,33 +1,29 @@
...
+                L = S*S
+                x = fill(0., SMatrix{S, S, T, L})
+                x * x
+                y = fill(0., SVector{S, T})
+                x * y
+                inv(x)
+                transpose(x)
+                _adjoint(Size(S, S), x)
+                # @assert precompile(Tuple{Core.kwftype(typeof(minimum)),NamedTuple{(:dims,), Tuple{Int}},typeof(minimum),SMatrix{S, S, T, L}})
+                # @assert precompile(Tuple{Core.kwftype(typeof(maximum)),NamedTuple{(:dims,), Tuple{Int}},typeof(maximum),SMatrix{S, S, T, L}})
```
Comment (timholy) [inline src/precompile.jl:14]:
```suggestion
minimum(x; dims=1); minimum(x; dims=2)
maximum(x; dims=1); maximum(x; dims=2)
```

File: src/precompile.jl [hunk 2]
```diff
@@ -1,33 +1,29 @@
...
-    end
 
-    # Some expensive generators
-    @assert precompile(Tuple{typeof(which(__broadcast,(Any,Size,Tuple{Vararg{Size}},Vararg{Any},)).generator.gen),Any,Any,Any,Any,Any,Any})
-    @assert precompile(Tuple{typeof(which(_zeros,(Size,Type{<:StaticArray},)).generator.gen),Any,Any,Any,Type,Any})
-    @assert precompile(Tuple{typeof(which(_mapfoldl,(Any,Any,Colon,Any,Size,Vararg{StaticArray},)).generator.gen),Any,Any,Any,Any,Any,Any,Any,Any})
+        # Some expensive generators
+        # @assert precompile(Tuple{typeof(which(__broadcast,(Any,Size,Tuple{Vararg{Size}},Vararg{Any},)).generator.gen),Any,Any,Any,Any,Any,Any})
+        # @assert precompile(Tuple{typeof(which(_zeros,(Size,Type{<:StaticArray},)).generator.gen),Any,Any,Any,Type,Any})
+        # @assert precompile(Tuple{typeof(which(_mapfoldl,(Any,Any,Colon,Any,Size,Vararg{StaticArray},)).generator.gen),Any,Any,Any,Any,Any,Any,Any,Any})
```
Comment (timholy) [inline src/precompile.jl:22]:
```suggestion
y .* x .* y'
zeros(y)
zeros(x)
```
I _think_ the `dims`-based `maximum`/`minimum` will cover the `mapfoldl`, not sure though. No biggie if not.

**Review by timholy — COMMENTED** [review]
Looks great, thanks!

**Comment by IanButterworth** [comment]
```
ERROR: LoadError: MethodError: no method matching zeros(::StaticArraysCore.SVector{2, Float32})

Closest candidates are:
  zeros(::Type{SA}) where {T, SA<:(StaticArraysCore.StaticArray{<:Tuple, T})}
   @ StaticArrays ~/work/StaticArrays.jl/StaticArrays.jl/src/arraymath.jl:2
  zeros(::Type{SA}) where SA<:(StaticArraysCore.StaticArray)
   @ StaticArrays ~/work/StaticArrays.jl/StaticArrays.jl/src/arraymath.jl:1
  zeros(::Type{T}, Union{Integer, AbstractUnitRange}...) where T
   @ Base array.jl:630
```

File: src/precompile.jl [hunk 3]
```diff
@@ -1,35 +1,29 @@
...
+                transpose(x)
+                _adjoint(Size(S, S), x)
+                minimum(x; dims=1)
+                minimum(x; dims=2)
+                maximum(x; dims=1)
+                maximum(x; dims=2)
+                getindex(x, SOneTo(S-1), SOneTo(S-1))
+                y .* x .* y'
+                zeros(y)
+                zeros(x)
```
Comment (IanButterworth) [inline src/precompile.jl:20]: @timholy see error above. I guessed

```suggestion
                zeros(T, y)
                zeros(T, x)
```
but hit 
```
ERROR: LoadError: MethodError: no method matching zeros(::Type{Float32}, ::StaticArraysCore.SVector{2, Float32})

Closest candidates are:
  zeros(::Type{T}, ::Union{Integer, AbstractUnitRange}...) where T
   @ Base array.jl:580
  zeros(::Type{T}, ::Tuple{}) where T
   @ Base array.jl:588
  zeros(::Type{T}, ::Tuple{Vararg{Integer, N}}) where {T, N}
   @ Base array.jl:583
  ...
```

Comment (timholy) [reply src/precompile.jl:20]: Shoot, should be `zero` not `zeros`. Sorry 'bout that!

**Review by timholy — APPROVED** [review]
In case I don't notice, someone should merge this when tests pass (nightly is now failing for a different reason having to do with nicer `@code_llvm` output; if one snips after the first `\n` I think that will fix that bug).

**Comment by KnutAM** [comment]
> In case I don't notice, someone should merge this when tests pass

I just scrolled by and figured I could give a ping: @timholy

**Comment by IanButterworth** [comment]
Can someone rerun nightly CI

**Comment by hyrodium** [comment]
The CI failure on nightly seems unrelated to this PR.

**Review by hyrodium — APPROVED** [review]
LGTM!

**Pull request merged**